### PR TITLE
bench get_retransmit_peers

### PR DESCRIPTION
**Pull request by jbiseda**
#### Problem

No bench for baseline get_retransmit_peers

#### Summary of Changes

Refactor to add benches for get_retransmit_peers_compat and get_retransmit_peers_deterministic

Fixes #

### CODE REVIEW

File: core/benches/cluster_nodes.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    let cluster_nodes = new_cluster_nodes::<RetransmitStage>(&cluster_info, &stakes);
+    (nodes, cluster_nodes)
+}
+
+#[bench]
+fn bench_get_retransmit_peers_deterministic(b: &mut Bencher) {
+    let mut rng = rand::thread_rng();
+    let (nodes, cluster_nodes) = make_cluster_nodes(&mut rng);
+
+    let mut shred_seed = [0u8; 32];
```
Comment (carllin) [inline core/benches/cluster_nodes.rs:26]: Does it make sense to generate a different seed on every iteration to test different behaviors?

File: core/src/cluster_nodes.rs [hunk 2]
```diff
@@ -462,22 +479,60 @@ impl From<Pubkey> for NodeId {
...
+    nodes.shuffle(rng);
+    let this_node = nodes[0].clone();
+    let mut stakes: HashMap<Pubkey, u64> = nodes
+        .iter()
+        .filter_map(|node| {
+            if rng.gen_ratio(1, 7) {
+                None // No stake for some of the nodes.
+            } else {
+                Some((node.id, rng.gen_range(0, 20)))
+            }
```
Comment (carllin) [inline core/src/cluster_nodes.rs:503]: Out of curiosity, how does this ratio of `unstaked:staked` nodes of `1:6` affect the performance of the shuffle? Does more or less unstaked/staked nodes make perf faster?

File: core/benches/cluster_nodes.rs (hunk 1 above)
Comment (jbiseda) [reply core/benches/cluster_nodes.rs]: I've changed this slightly to more accurately represent the seed generation for each case and to include multiple shreds in the calculation. I didn't want to include random generation of the inputs to `get_retransmit_peers_*` measurements.

File: core/src/cluster_nodes.rs (hunk 2 above)
Comment (jbiseda) [reply core/src/cluster_nodes.rs]: Added some tests for comparison:

```
running 4 tests
test bench_get_retransmit_peers_compat_unstaked_ratio_1_2         ... bench:   1,679,677 ns/iter (+/- 2,628)
test bench_get_retransmit_peers_compat_unstaked_ratio_1_32        ... bench:   1,690,616 ns/iter (+/- 3,048)
test bench_get_retransmit_peers_deterministic_unstaked_ratio_1_2  ... bench:   1,478,714 ns/iter (+/- 13,556)
test bench_get_retransmit_peers_deterministic_unstaked_ratio_1_32 ... bench:   2,476,904 ns/iter (+/- 39,032)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured; 0 filtered out; finished in 5.22s
```

File: core/src/cluster_nodes.rs [hunk 3]
```diff
@@ -39,7 +43,8 @@ enum NodeId {
     Pubkey(Pubkey),
 }
 
-struct Node {
+// pub for bench
```
Comment (behzadnouri) [inline core/src/cluster_nodes.rs:46]: If someone removes `pub`, the code will not compile; so the comment will be redundant.
Similarly other places below.

File: core/src/cluster_nodes.rs [hunk 4]
```diff
@@ -462,22 +479,62 @@ impl From<Pubkey> for NodeId {
     }
 }
 
+// pub for bench
+pub fn make_cluster<R: Rng>(
```
Comment (behzadnouri) [inline core/src/cluster_nodes.rs:483]: maybe rename this to `make_rand_cluster` to indicate this is a test/simulation only function.

**Review by behzadnouri — APPROVED** [review]
LGTM

File: core/src/cluster_nodes.rs (hunk 2 above)
Comment (carllin) [reply core/src/cluster_nodes.rs]: thanks for the insights!

**Review by carllin — APPROVED** [review]

File: core/src/cluster_nodes.rs (hunk 4 above)
Comment (jbiseda) [reply core/src/cluster_nodes.rs]: changed to `make_test_cluster`

**Pull request merged**